public/app/plugins/panel/flamegraph/components/FlameGraph/rendering.ts

```diff
@@ -8,10 +11,99 @@ import {
   LABEL_THRESHOLD,
   PIXELS_PER_LEVEL,
 } from '../../constants';
-import { TextAlign } from '../types';
+import { ClickedItemData, TextAlign } from '../types';
 
 import { FlameGraphDataContainer, LevelItem } from './dataTransform';
 
+const ufuzzy = new uFuzzy();
+
+export function useFlameRender(
+  canvasRef: RefObject<HTMLCanvasElement>,
+  data: FlameGraphDataContainer,
+  levels: LevelItem[][],
+  wrapperWidth: number,
+  rangeMin: number,
+  rangeMax: number,
+  search: string,
+  textAlign: TextAlign,
+  totalTicks: number,
+  focusedItemData?: ClickedItemData
+) {
+  const foundLabels = useMemo(() => {
+    if (search) {
+      const foundLabels = new Set<string>();
+      let idxs = ufuzzy.filter(data.getUniqueLabels(), search);
+
+      if (idxs) {
+        for (let idx of idxs) {
+          foundLabels.add(data.getUniqueLabels()[idx]);
+        }
+      }
+
+      return foundLabels;
+    }
+    // In this case undefined means there was no search so no attempt to highlighting anything should be made.
+    return undefined;
+  }, [search, data]);
+
+  const ctx = useSetupCanvas(canvasRef, wrapperWidth, levels.length);
+
+  useEffect(() => {
+    if (!ctx) {
+      return;
+    }
+    ctx.clearRect(0, 0, ctx.canvas.width, ctx.canvas.height);
+    const pixelsPerTick = (wrapperWidth * window.devicePixelRatio) / totalTicks / (rangeMax - rangeMin);
+
+    for (let levelIndex = 0; levelIndex < levels.length; levelIndex++) {
+      const level = levels[levelIndex];
+      // Get all the dimensions of the rectangles for the level. We do this by level instead of per rectangle, because
+      // sometimes we collapse multiple bars into single rect.
+      const dimensions = getRectDimensionsForLevel(data, level, levelIndex, totalTicks, rangeMin, pixelsPerTick);
+      for (const rect of dimensions) {
+        const focusedLevel = focusedItemData ? focusedItemData.level : 0;
+        // Render each rectangle based on the computed dimensions
+        renderRect(ctx, rect, totalTicks, rangeMin, rangeMax, levelIndex, focusedLevel, foundLabels, textAlign);
```

Comment:
We still render from the levels structure just some computations changed to count from the item.value instead of the value from data frame.